### PR TITLE
fix Kubectl describe node missing event

### PR DESCRIPTION
**Pull request by aimuz**
Kubectl describe node ignores the node events created by the controller

https://github.com/kubernetes/kubernetes/issues/106475

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by eddiezane** [comment]
Can you please add a test that helps make clear why we need this and why this is the correct fix?

**Comment by aimuz** [comment]
> Can you please add a test that helps make clear why we need this and why this is the correct fix?

Test cases have been added

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go [hunk 1]
```diff
@@ -4894,7 +4931,12 @@ func TestDescribeNode(t *testing.T) {
   memory             1Gi (8%)     2Gi (16%)
   ephemeral-storage  0 (0%)       0 (0%)
   hugepages-1Gi      0 (0%)       0 (0%)
-  hugepages-2Mi      512Mi (25%)  512Mi (25%)`}
+  hugepages-2Mi      512Mi (25%)  512Mi (25%)`, `Events:
+  Type    Reason  Age                      From  Message
+  ----    ------  ----                     ----  -------
+  Normal          7y342d                         Node bar status is now: NodeHasNoDiskPressure
+  Normal          7y342d (x2 over 7y342d)        Node bar status is now: NodeReady
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/describe/describe_test.go:4938]: This test now fails because the date is different.

**Review by eddiezane — COMMENTED** [review]
Thanks for adding a test. I apologize but I still don't understand this change. Can you please help me understand?

**Comment by aimuz** [comment]
> Can you please help me understand?

@eddiezane 

This test case is to test that both `nodename` as `event uid` and `node.uid `as `event uid `can be displayed by `kubectl describe node`.

I will fix this test case so that it can be tested at any time

**Comment by derekwaynecarr** [comment]
I have trouble understanding the problem being solved and the impact of the fix.

Is it easy to demonstrate via a test case?

**Comment by pacoxu** [comment]
#100847 tried to fix the problem in kubelet side #100236.
#100236 is clearer for the problem that @aimuz  met.

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go (hunk 1 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/describe/describe_test.go]: > This test now fails because the date is different.

I reworked the test cases

**Comment by aimuz** [comment]
@derekwaynecarr 

> #100847 tried to fix the problem in kubelet side #100236.
> #100236 is clearer for the problem that @aimuz met.

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go [hunk 2]
```diff
@@ -3523,7 +3523,10 @@ func (d *NodeDescriber) Describe(namespace, name string, describerSettings Descr
 			klog.Errorf("Unable to construct reference to '%#v': %v", node, err)
 		} else {
 			// TODO: We haven't decided the namespace for Node object yet.
-			ref.UID = types.UID(ref.Name)
+			// there are two UIDs for host events:
+			// controller use node.uid
+			// kubelet use node.name
+			ref.UID = ""
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/describe/describe.go:3529]: If the discrepancy is in the kubelet which is injecting name into UID that should be fixed on the kubelet side of things. What I'm willing to accept on the kubectl side of things are 2 `searchEvents` calls one using UID and other using name such that we'll scrape both events and leave a TODO here pointing to issue against kubelet, such that we'll be able to remove the workaround after kubelet is fixed.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go (hunk 2 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/describe/describe.go:3529]: @soltysh 
I have a question, by what field are the events sorted, as the merge needs to combine them and reorder them, is it the creation time field?

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go [hunk 3]
```diff
@@ -3533,8 +3533,17 @@ func (d *NodeDescriber) Describe(namespace, name string, describerSettings Descr
 			klog.Errorf("Unable to construct reference to '%#v': %v", node, err)
 		} else {
 			// TODO: We haven't decided the namespace for Node object yet.
-			ref.UID = types.UID(ref.Name)
+			// there are two UIDs for host events:
+			// controller use node.uid
+			// kubelet use node.name
+			// TODO: Uniform use of UID
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/describe/describe.go:3539]: Nit: would be nice to point to a specific issue number tracking this problem.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/triage accepted
/priority backlog

**Pull request merged**